### ISSUE TITLE
Session.ScheduleTask and problems when task is not scheduled, but executed directly

### ISSUE THREAD

**Issue by chrhol** [Starcounter/Starcounter#3529]
In  a couple of different issues now we found a problem with the current implementation of `Session.ScheduleTask` and that is that if the session that is scheduled belongs to the same scheduler as the current running one, the task is not scheduled, but executed directly and there is no way for the user to control this.

In #3522 this might lead to a deadlock in the administrator, due to parts of the administrator has it's own locks and in #3490 the current transaction is locked on the thread, so accessing a viewmodel attached with another transaction will fail in the case when the task is executed directly.

We need to change this API so it either always schedules the task or add a possiblity to force a task to be scheduled.

/cc @alemoi

**Comment by alemoi** [comment]
Just curious, is it possible to do the transactions switch and still execute tasks on the same scheduler? If we would hypothetically have no `Session.ScheduleTask`, and have something to do on a different transaction/session. How would we do that? @chrhol can you elaborate on that, since you know how transactions are switched I suppose. Let's try to find the right way first, and adopt `Session.ScheduleTask` and `TaskScheduling` to it.

**Comment by chrhol** [comment]
Well I don't think in the issues that were referenced that it would be possible. When called from commithooks and as the administrator is implemented with multiple threads accessing it, async access to sessions is required I think. 

Actually, I think `Session.Use` should maybe not be public either and when you need to do something in a session you have to schedule a task for it since it needs to be run on the correct scheduler.

**Comment by alemoi** [comment]
I see the only problem is sync calls in ScheduleTask. Otherwise we would be able to schedule the task always and forget about it. @chrhol do you confirm that in all cases above it only about sync calls?

**Comment by chrhol** [comment]
> @chrhol do you confirm that in all cases above it only about sync calls? (ScheduleTask with waitForCompletion true)

Yes, the problem is sync calls, but not only when `waitForCompletion` is `true`, since we make a shortcut, even if it is set to `false`, if we already are on the correct scheduler.

**Comment by chrhol** [comment]
Maybe it's enough to tweak so that if waitForCompletion is false we always schedule a task?

**Comment by alemoi** [comment]
> Yes, the problem is sync calls, but not only when waitForCompletion is true, since we make a shortcut, even if it is set to false, if we already are on the correct scheduler.

This I can make so it schedules always. For waiting for completion on the same scheduler we might use some reset count event for synchronization or try to use existing code. Not sure if that will work on the same scheduler, I remember there is some issue with possible deadlock. @eriohl maybe you remember? Thanks.

**Comment by alemoi** [comment]
> Maybe it's enough to tweak so that if waitForCompletion is false we always schedule a task?

I thought to suggest this, but then automatically thought on how users will do the manual waiting for the task to complete on the same scheduler. I'm just worrying about the possible deadlock. That I am asking above.

**Comment by eriohl** [comment]
If you block a scheduler waiting for another task on the same scheduler this results in a deadlock. Same could apply if you are waiting for another scheduler since this scheduler might be waiting for you. Thus you may only block a scheduler if what you are waiting for can not possibly wait for any other scheduler directly or indirectly. This is why wait and retry on queue full is kind of risky. But you can block the current task and allow the scheduler to run another.

**Comment by alemoi** [comment]
Thanks for the response @eriohl 
So basically if I schedule a task on the current scheduler and try to use manual reset event to wait for the event in this task to set when the scheduled task is run - then it will result in a deadlock. Is that correct?

I am just trying to figure out how to wait in one task for the other task to finish.

**Comment by eriohl** [comment]
> So basically if I schedule a task on the current scheduler and try to use manual reset event to wait for the event in this task to set when the scheduled task is run - then it will result in a deadlock. Is that correct?

If you use a non-alertable wait then you will block the scheduler. If you use an alertable wait (like the managed manual reset event) the block detection will catch this. Quite a bit of overhead however (which is why the old implemention didn't use this). Exception perhaps being if you run in the thread cap. Not entirely sure this is handled now.

**Comment by alemoi** [comment]
I have tested the performance difference between manual reset event and RunSync on a different scheduler. The difference is significant I would say, especially when it comes to running and waiting on the same scheduler:
```
##teamcity[buildStatisticValue key='SimplestSyncTaskOnSpecificScheduler1_PerSec' value='53705']
##teamcity[buildStatisticValue key='SimplestAsyncWithManualResetEventSameScheduler_PerSec' value='36509']
##teamcity[buildStatisticValue key='SimplestAsyncWithManualResetEventScheduler1_PerSec' value='41203']
```

**Comment by alemoi** [comment]
Regarding RunSync I think we have no choice but using manual reset events or some alternative. Yes, the performance is different, but in any case very low i.e. not comparable to normal handlers processing.

**Comment by alemoi** [comment]
I have changed the code now for RunSync to always schedule a task and wait for it using manual reset event. This has worse performance, but previous RunSync implementation was anyway slow. As I remember there were a couple of issues related to this. CC @chrhol @andwah to check if previous problems were solved using this patch.

**Comment by chrhol** [comment]
> I have changed the code now for RunSync to always schedule a task and wait for it using manual reset event. This has worse performance, but previous RunSync implementation was anyway slow. As I remember there were a couple of issues related to this. CC @chrhol @andwah to check if previous problems were solved using this patch.

I think we need to fix the code in `Session.ScheduleTask` as well, since it bypasses the scheduling sometimes and executes the task directly. It should probably always call `Scheduling.ScheduleTask`

**Comment by alemoi** [comment]
This one I forgot, that it may run directly skipping scheduling. I will check it. Thank you.

**Comment by chrhol** [comment]
I think the things (i.e hacks :)) I added can be removed if this is changed. They shouldn't be needed if the task is always scheduled. 

There is one case when there will be a timeout though with this. If `waitForCompletion` is `true` and `Session.ScheduleTask` is called for the same session as currently is active, the scheduled task will never get access to the session, which will lead to an exception.

Maybe this would be considered an incorrect usage from usercode though, since it would be easy to just check the current session in the usercode.

**Comment by alemoi** [comment]
Thanks for the input! Now the things are pushed so you and @andwah can test, if previous problems are resolved.

**Comment by andwah** [comment]
I just pulled the latest and i will do some testing during the day.

**Comment by chrhol** [comment]
Nice! :+1: 

I've done some testing using Warehouse/PlaygroundKeeper and can no longer reproduce the deadlock.

Will test a bit more and also try it from a commithook, but so far it looks like it's working fine.

**Comment by andwah** [comment]
I have not got any deadlocks during a full day work, usually i get a couple of them.

Good work!

**Comment by chrhol** [comment]
I haven't got any deadlocks either and I also tested scheduling from commithooks and that works fine. I think we can consider this issue resolved!
